packages/jest-core/src/SearchSource.ts

```diff
@@ -37,8 +37,57 @@ export type TestSelectionConfig = {
   watch?: boolean;
 };
 
-const globsToMatcher = (globs: Array<Config.Glob>) => (path: Config.Path) =>
-  micromatch([replacePathSepForGlob(path)], globs, {dot: true}).length > 0;
+const globsMatchers = new Map<
+  string,
+  {
+    isMatch: (str: string) => boolean;
+    negated: boolean;
+  }
+>();
+
+const globsToMatcher = (globs: Array<Config.Glob>) => {
+  const matchers = globs.map(glob => {
+    if (!globsMatchers.has(glob)) {
+      const state = micromatch.scan(glob, {dot: true});
+      const matcher = {
+        isMatch: micromatch.matcher(glob, {dot: true}),
+        negated: state.negated,
+      };
+      globsMatchers.set(glob, matcher);
+    }
+    return globsMatchers.get(glob);
+  });
+
+  return (path: Config.Path) => {
+    const omit = new Set();
+    const keep = new Set();
```

Comment:
I think I can probably simplify this to be just strings instead of Sets, because we are only ever dealing with a single path here.